Update Skia to a more recent version (December 2016)

livecodefraser opened the pull request:
Currently a work-in-progress. Depends on using libc++ on macOS as Skia now uses modern C++ features.

Per-platform status:

- MacOS: done
- iOS: done
- Android: done
- Emscripten: done
- Windows: in progress
- Linux: in progress

There is a corresponding PR for this in the third-party repo; these are just the source code changes required to adapt to the changes in Skia's API.

Various bits are still broken. In particular, legacy blend modes have been removed from libgraphics (though incompletely; assertion failures will be triggered when used). It is not possible to re-add legacy blend modes due to major changes in how Skia handles blending. Additionally, the gradient code (in legacygradients.cpp) does not compile due to a change in the Skia filtering API; this will need to be refactored (probably to use Skia's built-in gradient support just like the widget graphics library provides).

livecodefraser committed:
Revert "Revert "build: Compile for deployment on Mac OS X 10.9 with libc++""

This reverts commit 2c92018d5d3ff6b52f435713eb2c3b1b831b69f6.
Using libc++ is a requirement for building recent versions of Skia.

livecodefraser committed:
Update libgraphics to account for Skia API changes

The biggest impact is the removal of custom blend modes in Skia. This
forces us to drop support for legacy blend modes as we can no longer
emulate them.

On the plus side, only using the built-in Skia blend modes will result
in faster blending.

livecodefraser commented:
Update in thirdparty: https://github.com/livecode/livecode-thirdparty/pull/92

peter-b commented:
I have tested this in the `integration-vs2015` branch.

<details><summary>Errors in Android build</summary>

[Full log](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1039/steps/compile/logs/stdio)

```
  CXX(target) out/Fast/obj.target/libskia/thirdparty/libskia/src/codec/SkSwizzler.o
In file included from ../../thirdparty/libskia/src/codec/SkSwizzler.cpp:11:
In file included from ../../thirdparty/libskia/src/core/SkOpts.h:13:
In file included from ../../thirdparty/libskia/src/utils/SkTextureCompressor.h:11:
In file included from ../../thirdparty/libskia/src/core/SkBitmapProcShader.h:10:
In file included from ../../thirdparty/libskia/src/core/SkImagePriv.h:12:
../../thirdparty/libskia/src/core/SkSmallAllocator.h:30:29: error: unknown type name 'max_align_t'
    using SystemAlignment = max_align_t;
                            ^
../../thirdparty/libskia/src/core/SkSmallAllocator.h:109:50: error: use of undeclared identifier 'SystemAlignment'
    static constexpr size_t kAlignment = alignof(SystemAlignment);
                                                 ^
2 errors generated.
```

</details>

<details><summary>Errors in Emscripten build</summary>

[Full log](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1045/steps/compile/logs/stdio)

```
  CXX(target) out/Fast/obj.target/libskia/thirdparty/libskia/src/codec/SkSwizzler.o
WARNING:root:not all asm.js optimizations are possible with ALLOW_MEMORY_GROWTH, disabling those
In file included from ../../thirdparty/libskia/src/codec/SkSwizzler.cpp:11:
In file included from ../../thirdparty/libskia/src/core/SkOpts.h:13:
In file included from ../../thirdparty/libskia/src/utils/SkTextureCompressor.h:11:
In file included from ../../thirdparty/libskia/src/core/SkBitmapProcShader.h:10:
In file included from ../../thirdparty/libskia/src/core/SkImagePriv.h:12:
../../thirdparty/libskia/src/core/SkSmallAllocator.h:30:29: error: unknown type name 'max_align_t'; did you mean 'std::max_align_t'?
    using SystemAlignment = max_align_t;
                            ^~~~~~~~~~~
                            std::max_align_t
/toolchain/emscripten-js/emsdk_portable/emscripten/master/system/include/libcxx/cstddef:53:21: note: 'std::max_align_t' declared here
typedef long double max_align_t;
                    ^
1 error generated.
ERROR:root:compiler frontend failed to generate LLVM bitcode, halting
thirdparty/libskia/libskia.target.mk:982: recipe for target 'out/Fast/obj.target/libskia/thirdparty/libskia/src/codec/SkSwizzler.o' failed
```

</details>

<details><summary>Errors in x86/x86_64 Linux build</summary>

[Full linux-x86 log](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1044/steps/compile/logs/stdio)
[Full linux-x86_64 log](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1039/steps/compile/logs/stdio)

```
  CXX(target) out/Fast/obj.target/libGraphics/libgraphics/src/context.o
In file included from ../../thirdparty/libskia/include/core/../private/SkMutex.h:11:0,
                 from ../../thirdparty/libskia/include/core/SkImageFilter.h:13,
                 from ../../libgraphics/src/context.cpp:29:
../../thirdparty/libskia/include/core/../private/../private/SkSemaphore.h:53:28: error: 'SkBaseSemaphore::SkBaseSemaphore' names constructor
In file included from ../../thirdparty/libskia/include/core/SkImageFilter.h:13:0,
                 from ../../libgraphics/src/context.cpp:29:
../../thirdparty/libskia/include/core/../private/SkMutex.h:43:24: error: 'SkBaseMutex::SkBaseMutex' names constructor
../../libgraphics/src/context.cpp:429:13: warning: 'bool MCGContextCreateWithSurface(sk_sp<SkSurface>, __MCGContext*&)' defined but not used [-Wunused-function]
```

</details>

<details><summary>Errors in Mac build</summary>

[Full log](https://vulcan.livecode.com/builders/try-community-mac/builds/1047/steps/compile/logs/stdio)

```
CompileC /Users/administrator/Vulcan/slaves/build-mac-3-1/slave/try-community-mac/worktree/_cache/mac/Fast/libskia.build/Objects-normal/i386/SkBitmapProcState_opts_SSSE3.o src/opts/SkBitmapProcState_opts_SSSE3.cpp normal i386 c++ com.apple.compilers.llvm.clang.1_0.compiler
    cd /Users/administrator/Vulcan/slaves/build-mac-3-1/slave/try-community-mac/worktree/thirdparty/libskia
    export LANG=en_US.US-ASCII
    /Users/administrator/Vulcan/toolchain/mac/Xcode_8_2_1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -x c++ -arch i386 -fmessage-length=0 -fdiagnostics-show-note-include-stack -fmacro-backtrace-limit=0 -std=c++11 -stdlib=libc++ -Wno-trigraphs -fno-exceptions -fno-rtti -fpascal-strings -O0 -Wno-missing-field-initializers -Wno-missing-prototypes -Wno-return-type -Wno-non-virtual-dtor -Wno-overloaded-virtual -Wno-exit-time-destructors -Wno-missing-braces -Wparentheses -Wswitch -Wno-unused-function -Wno-unused-label -Wno-unused-parameter -Wno-unused-variable -Wunused-value -Wno-empty-body -Wno-uninitialized -Wno-unknown-pragmas -Wno-shadow -Wno-four-char-constants -Wno-conversion -Wno-constant-conversion -Wno-int-conversion -Wno-bool-conversion -Wno-enum-conversion -Wno-shorten-64-to-32 -Wno-newline-eof -Wno-c++11-extensions -DSK_RELEASE -DSK_IGNORE_ETC1_SUPPORT -DSK_SUPPORT_LEGACY_IMAGE_ENCODER_CLASS -DSK_SUPPORT_LEGACY_CANVAS_IS_REFCNT -DSK_SUPPORT_LEGACY_GETTOPDEVICE -DSK_SUPPORT_LEGACY_ACCESSBITMAP -DSK_SUPPORT_LEGACY_CLIP_REGIONOPS -DSK_SUPPORT_LEGACY_GETDEVICE -DTARGET_PLATFORM_MACOS_X -D_MACOSX -D_RELEASE -DNDEBUG -isysroot /Users/administrator/Vulcan/toolchain/mac/Xcode_8_2_1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk -fasm-blocks -fstrict-aliasing -Wno-deprecated-declarations -Winvalid-offsetof -mmacosx-version-min=10.9 -g -fvisibility=hidden -fno-threadsafe-statics -Wno-sign-conversion -Wno-infinite-recursion -Wno-move -iquote /Users/administrator/Vulcan/slaves/build-mac-3-1/slave/try-community-mac/worktree/_cache/mac/Fast/libskia.build/skia-generated-files.hmap -I/Users/administrator/Vulcan/slaves/build-mac-3-1/slave/try-community-mac/worktree/_cache/mac/Fast/libskia.build/skia-own-target-headers.hmap -I/Users/administrator/Vulcan/slaves/build-mac-3-1/slave/try-community-mac/worktree/_cache/mac/Fast/libskia.build/skia-all-target-headers.hmap -iquote /Users/administrator/Vulcan/slaves/build-mac-3-1/slave/try-community-mac/worktree/_cache/mac/Fast/libskia.build/skia-project-headers.hmap -I/Users/administrator/Vulcan/slaves/build-mac-3-1/slave/try-community-mac/worktree/_build/mac/Fast/include -Iinclude/android -Iinclude/animator -Iinclude/c -Iinclude/codec -Iinclude/config -Iinclude/core -Iinclude/effects -Iinclude/gpu -Iinclude/images -Iinclude/pathops -Iinclude/ports -Iinclude/private -Iinclude/svg -Iinclude/utils -Iinclude/utils/ios -Iinclude/utils/mac -Iinclude/utils/win -Iinclude/views -Iinclude/xml -Isrc/android -Isrc/animator -Isrc/c -Isrc/codec -Isrc/core -Isrc/effects -Isrc/fonts -Isrc/gpu -Isrc/image -Isrc/images -Isrc/lazy -Isrc/opts -Isrc/pathops -Isrc/pdf -Isrc/pipe -Isrc/ports -Isrc/sfnt -Isrc/sksl -Isrc/svg -Isrc/utils -Isrc/views -Isrc/xml -Isrc/xps -I../libgif/include -I../libexpat -I../libexpat/lib -I../libjpeg/include -I../libpng/include -I../libz/include -I/Users/administrator/Vulcan/slaves/build-mac-3-1/slave/try-community-mac/worktree/_cache/mac/Fast/libskia.build/DerivedSources/i386 -I/Users/administrator/Vulcan/slaves/build-mac-3-1/slave/try-community-mac/worktree/_cache/mac/Fast/libskia.build/DerivedSources -w -F/Users/administrator/Vulcan/slaves/build-mac-3-1/slave/try-community-mac/worktree/_build/mac/Fast -MMD -MT dependencies -MF /Users/administrator/Vulcan/slaves/build-mac-3-1/slave/try-community-mac/worktree/_cache/mac/Fast/libskia.build/Objects-normal/i386/SkBitmapProcState_opts_SSSE3.d --serialize-diagnostics /Users/administrator/Vulcan/slaves/build-mac-3-1/slave/try-community-mac/worktree/_cache/mac/Fast/libskia.build/Objects-normal/i386/SkBitmapProcState_opts_SSSE3.dia -c /Users/administrator/Vulcan/slaves/build-mac-3-1/slave/try-community-mac/worktree/thirdparty/libskia/src/opts/SkBitmapProcState_opts_SSSE3.cpp -o /Users/administrator/Vulcan/slaves/build-mac-3-1/slave/try-community-mac/worktree/_cache/mac/Fast/libskia.build/Objects-normal/i386/SkBitmapProcState_opts_SSSE3.o
/Users/administrator/Vulcan/slaves/build-mac-3-1/slave/try-community-mac/worktree/thirdparty/libskia/src/opts/SkBitmapProcState_opts_SSSE3.cpp:66:13: error: always_inline function '_mm_shuffle_epi8' requires target feature 'ssse3', but would be inlined into function 'PrepareConstantsTwoPixelPairs' that is compiled without support for 'ssse3'
    all_x = _mm_shuffle_epi8(all_x, mask_dist_select);
            ^
/Users/administrator/Vulcan/slaves/build-mac-3-1/slave/try-community-mac/worktree/thirdparty/libskia/src/opts/SkBitmapProcState_opts_SSSE3.cpp:161:12: error: always_inline function '_mm_maddubs_epi16' requires target feature 'ssse3', but would be inlined into function 'ProcessPixelPairHelper' that is compiled without support for 'ssse3'
    return _mm_maddubs_epi16(a0, scale_x);
           ^
/Users/administrator/Vulcan/slaves/build-mac-3-1/slave/try-community-mac/worktree/thirdparty/libskia/src/opts/SkBitmapProcState_opts_SSSE3.cpp:229:19: error: always_inline function '_mm_maddubs_epi16' requires target feature 'ssse3', but would be inlined into function 'ProcessOnePixelZeroSubY' that is compiled without support for 'ssse3'
    __m128i sum = _mm_maddubs_epi16(a0, scale_x);
                  ^
/Users/administrator/Vulcan/slaves/build-mac-3-1/slave/try-community-mac/worktree/thirdparty/libskia/src/opts/SkBitmapProcState_opts_SSSE3.cpp:355:10: error: always_inline function '_mm_maddubs_epi16' requires target feature 'ssse3', but would be inlined into function 'ProcessOnePixel' that is compiled without support for 'ssse3'
    a0 = _mm_maddubs_epi16(a0, scale_x);
         ^
/Users/administrator/Vulcan/slaves/build-mac-3-1/slave/try-community-mac/worktree/thirdparty/libskia/src/opts/SkBitmapProcState_opts_SSSE3.cpp:229:19: error: always_inline function '_mm_maddubs_epi16' requires target feature 'ssse3', but would be inlined into function 'ProcessOnePixelZeroSubY' that is compiled without support for 'ssse3'
    __m128i sum = _mm_maddubs_epi16(a0, scale_x);
                  ^
/Users/administrator/Vulcan/slaves/build-mac-3-1/slave/try-community-mac/worktree/thirdparty/libskia/src/opts/SkBitmapProcState_opts_SSSE3.cpp:117:14: error: always_inline function '_mm_shuffle_epi8' requires target feature 'ssse3', but would be inlined into function 'PrepareConstantsTwoPixelPairsDXDY' that is compiled without support for 'ssse3'
    all_xy = _mm_shuffle_epi8(all_xy, mask_dist_select);
             ^
6 errors generated.
```

</details>

<details><summary>Errors in iPhoneSimulator 10.2 build</summary>

[Full log](https://vulcan.livecode.com/builders/try-community-ios/builds/1049/steps/compile-iphonesimulator10.2/logs/stdio)

```
ld: scattered reloc r_address too large for architecture i386
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
</details>

<details><summary>Errors in x86 Windows build</summary>

[Full log](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1061/steps/compile/logs/stdio)

Many errors compiling `libskia\src\xps\SkXPSDevice.cpp`; see log.

</details>

livecodefraser commented:
- Android: the toolchain on the build server is too old (easy to update, at least)
- Emscripten: needs some thinking about; will need to tweak Skia config macros...
- Linux: toolchain is too old 😢 (it's rejecting a valid C++11 construct)
- macOS: fixed
- iPhoneSimulator: fixed (by dropping 32-bit simulator support)
- Windows: fixed

livecodefraser commented:
I've now added a couple of patches to Skia to (hopefully) avoid the missing C++11 feature that is giving us problems on Linux and to correct the Skia feature detection for Emscripten. (All changes are in the thirdparty repo).

livecodeali commented on engine/src/globdefs.h:
```diff
@@ -53,20 +53,6 @@ along with LiveCode.  If not see <http://www.gnu.org/licenses/>.  */
-#undef _WINDOWS
-
-#endif
-
```
This section needs to be reinstated to have things working at the moment.
